network/limitlistener/rejectingLimitListener.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// RejectingLimitListener is a modification of LimitListener in
+// "golang.org/x/net/netutil". The difference is that when the number of connections
+// exceeds the limit, RejectingLimitListener will accept and immediately close all
+// new connections.
+
+package limitlistener
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+)
+
+// RejectingLimitListener returns a Listener that accepts at most n simultaneous
+// connections from the provided Listener.
+func RejectingLimitListener(l net.Listener, n uint64) net.Listener {
+	return &rejectingLimitListener{
+		Listener: l,
+		sem:      make(chan struct{}, n),
+		done:     make(chan struct{}),
+	}
+}
+
+type rejectingLimitListener struct {
+	net.Listener
+	sem       chan struct{}
+	closeOnce sync.Once     // ensures the done chan is only closed once
+	done      chan struct{} // no values sent; closed when Close is called
+}
+
+func (l *rejectingLimitListener) release() {
+	<-l.sem
+}
+
+func (l *rejectingLimitListener) Accept() (net.Conn, error) {
+	for {
+		select {
+		case <-l.done:
+			return nil, errors.New("Accept() limit listener is closed")
+		default:
+			c, err := l.Listener.Accept()
+			if err != nil {
+				return nil, fmt.Errorf("Accept() accept err: %w", err)
+			}
+			select {
+			case l.sem <- struct{}{}:
+				return &rejectingLimitListenerConn{Conn: c, release: l.release}, nil
+			default:
+				// Close connection immediately.
+				err = c.Close()
+				if err != nil {
+					return nil, fmt.Errorf("Accept() close err: %w", err)
```

Comment:
while I agree that we want to log this case, I think that we shouldn't return an error here.
returning an error here would be an attack vector : if someone could cause the Close here to return an error, it would exit from the Listening goroutine, and would no longer receive new connections.

Reply:
yes; the alternative is to catch this error at the caller, and then to treat it differently. But I'm not sure that it's an easier option.

Reply:
done